examples/asset/loading_screen.rs

```diff
@@ -0,0 +1,230 @@
+use bevy::{gltf::Gltf, prelude::*};
+
+/// This example illustrates, how a loading screen can be implemented. It has an animated spinner and listens for an `AssetEvent`.
+
+#[derive(Clone, Eq, PartialEq, Debug, Hash)]
+enum GameState {
+    Loading,
+    Playing,
+}
+
+fn main() {
+    App::new()
+        .insert_resource(Msaa { samples: 4 })
+        .add_plugins(DefaultPlugins)
+        // Loading screen plugin and its `asset_listening_system`
+        .add_plugin(plugin::SimpleLoadingScreenPlugin {
+            loading_state: GameState::Loading,
```

Comment:
I'm not sure this needs to be internal state to the plugin. It's not really use anywhere inside of it. this could just be a state on the App.

Reply:
Or, I think it's possible to move the asset_listening system inside the plugin. Let the plugin insert a LoadingState and have a system in the main example that simply listens to that State change

Reply:
> It's not really use anywhere inside of it.

It is used on line 117 and 120, so that the plugin "knows" at which state it has to work and when it needs to be closed. 🙂
https://github.com/bevyengine/bevy/blob/d36ceeffe3ed515a525667dfbb0e372d94a99802/examples/asset/loading_screen.rs#L117-L120

> Or, I think it's possible to move the asset_listening system inside the plugin.

My initial intention for this example was, to later extract the plugin into a lib in the example folder, to share it with other examples. If something like shared example code is ok within Bevy. That would mean, that the plugin doesn't know to which `AssetEvent`s it has to listen. That is the reason why i left the `asset_listening_system` in the main example. I wanted the plugin to take a `Fn` and store it internal, but the `Plugin` trait requires `Send + Sync` and the `Fn` would also need to be `'static` which would make it really ugly. And using a generic with `IntoSystemDescriptor<Params>` bound would also not work, because it isn't `Copy` or `Clone` and as such can't be used as param for `with_system`, because we only get it as borrowed reference inside `build`. 😅

That said...if there is no chance, that this will happen, then there is no reason to not move `asset_listening_system` inside the plugin.
Regarding `GameState`...i thought, that we need a "global" state to make sure, that the systems run in the correct order and didn't want to impose a state to a game/app that uses the plugin.

Reply:
For the state, I'd rather have a LoadingState that is controlled by the plugin entirely so that examples that need loading screens could simply do `.add_system_set(SystemSet::on_enter(LoadingState::Done).with_system(game_system))`. I believe states are already global by default if the struct holding the state is pub.

For the asset event issue, I didn't think about the fact that it was for specific asset types, but if it's only the gltf assets that are slow to load I wouldn't having it only for gltf. Or maybe a different plugin for each type. 

Essentially, I would prefer if examples had as little code as possible related to loading assets and with the current approach it would be to verbose to use in other examples. It's still a decent example on it's own to showcase how a loading screen would work.

Reply:
Hm...i'm not that deep into the stages and state stuff. What happens if an example, like `alien_cake_addict`, already has a state, can they be used together, is there a way to say, run `LoadingState` stuff first and afterwards run the `GameState` stuff?

Yeah...`AssetEvents` really are the problematic part in making this generic, without manually tracking (somehow registering) assets and their events in the plugin.